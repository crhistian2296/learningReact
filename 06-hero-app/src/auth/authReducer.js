// const user = {
//   name:'Fernando',
//   logged: true
// }

import { login, logout } from '../types/type';

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case login:
      return {
        ...action.payload,
        logged: true,
      };
    case logout:
      return {
        logged: false,
      };
    default:
      return state;
  }
};
