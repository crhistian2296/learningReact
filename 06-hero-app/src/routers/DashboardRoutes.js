import { Route, Routes } from 'react-router-dom';
// Pages
import DCScreen from '../components/dc/DCScreen';
import HeroScreen from '../components/heroes/HeroScreen';
import MarvelScreen from '../components/marvel/MarvelScreen';
import SearchScreen from '../components/search/SearchScreen';
// UI
import { Navbar } from '../components/UI/Navbar';

export const DashboardRoutes = () => {
  return (
    <>
      <Navbar />
      <div className='container my-3'>
        <Routes>
          <Route path='hero/:heroId' element={<HeroScreen />} />
          <Route path='marvel' element={<MarvelScreen />} />
          <Route path='dc' element={<DCScreen />} />
          <Route path='search' element={<SearchScreen />} />

          {/* Default */}
          <Route path='/' element={<MarvelScreen />} />
          {/* <Route path='*' element={<Navigate to='/' />} /> */}
        </Routes>
      </div>
    </>
  );

  // *react router v6 convention
  // return (
  //   <>
  //     <Navbar />
  //     <Outlet />
  //   </>
  // );
};
