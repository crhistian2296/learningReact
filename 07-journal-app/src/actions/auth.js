import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { uiFinishLoading, uiStartLoading } from './ui';

// export const startLoginEmailPassword = (email, password) => {
//   // *redux-thunk provee automaticamente ese dispatch para acciones asincronas
//   return (dispatch) => {
//     setTimeout(() => {
//       dispatch(login(123, 'crhistian'));
//     }, 3000);
//   };
// };

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
/**
 * Crea una cuenta en la app de autenticacion en firebase mediante correo
 * @param {string} email
 * @param {string} password
 * @param {string} name
 * @returns void
 */
export const startRegisterWithEmailAsync = (email, password, name) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};
export const startLoginWithEmailAsync = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};

/**
 * Crean una cuenta en la app de autenticacion de firebase mediante boton de google
 * @returns void
 */
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch(console.log);
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    await dispatch(logout());
  };
};

export const logout = () => ({ type: types.logout });
