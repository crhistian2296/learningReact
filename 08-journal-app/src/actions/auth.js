import { login } from '../types/types';

export const loginAction = (uid, displayName) => ({
  type: login,
  payload: {
    uid,
    displayName
  }
})
