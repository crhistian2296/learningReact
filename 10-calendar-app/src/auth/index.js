export * from './pages/Login.jsx';
