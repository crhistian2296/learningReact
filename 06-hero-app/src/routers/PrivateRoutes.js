import { useContext } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { AuthContext } from '../auth/AuthContext';

export const PrivateRoutes = ({ children }) => {
  const { user } = useContext(AuthContext);

  const { pathname, search } = useLocation();
  const lastUrl = pathname + search;

  localStorage.setItem('lastPath', lastUrl);

  return user.logged ? children : <Navigate to='/login' />;

  // * React router v6
  // return user.logged ? <Outlet/> : <Navigate to='/login' />
};
