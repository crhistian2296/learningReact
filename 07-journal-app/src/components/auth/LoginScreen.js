import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import useForm from '../../hooks/useFrom';

const LoginScreen = () => {
  const { formValues, handleInputChange } = useForm({
    email: 'cristof@gmail.com',
    password: '1234',
  });

  const { email, password } = formValues;

  const dispatch = useDispatch();

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(123456, 'cristof'));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className='box-container__title'>Login</h3>

      <form onSubmit={handleLogin}>
        <input
          type='text'
          placeholder='Email'
          name='email'
          className='box-container__input'
          value={email}
          onChange={handleInputChange}
          autoComplete='off'
        />
        <input
          type='password'
          placeholder='Password'
          name='password'
          className='box-container__input'
          value={password}
          onChange={handleInputChange}
        />
        <button type='submit' className='btn btn-primary btn-block mt-1'>
          Login
        </button>
        <div className='box-container__social-networks'>
          <p>Login with social networks</p>

          <div className='google-btn' onClick={handleGoogleLogin}>
            <div className='google-icon-wrapper'>
              <img
                className='google-icon'
                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                alt='google button'
              />
            </div>
            <p className='google-btn-text'>
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link to='/auth/register' className='link'>
          Create new account
        </Link>
      </form>
    </>
  );
};

export default LoginScreen;
