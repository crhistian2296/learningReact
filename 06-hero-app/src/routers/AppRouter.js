import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
import LoginScreen from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoutes } from './PrivateRoutes';
import { PublicRoutes } from './PublicRoutes';

const AppRouter = () => {
  return (
    <Router>
      <div className='main'>
        <Routes>
          {/* <Route path='login' element={<LoginScreen />} /> */}
          <Route
            path='login'
            element={
              <PublicRoutes>
                <LoginScreen />
              </PublicRoutes>
            }
          />
          <Route
            path='*'
            element={
              <PrivateRoutes>
                <DashboardRoutes />
              </PrivateRoutes>
            }
          />
        </Routes>
      </div>
    </Router>
  );

  // *react router v6 convention
  // return (
  //   <Router>
  //     <div className='main'>
  //       <Routes>
  //         <Route path='login' element={<LoginScreen />} />
  //         <Route path='' element={<PrivateRoutes />}>
  //           <Route element={<DashboardRoutes />}>
  //             <Route path='hero/:heroId' element={<HeroScreen />} />
  //             <Route path='marvel' element={<MarvelScreen />} />
  //             <Route path='dc' element={<DCScreen />} />
  //             <Route path='search' element={<SearchScreen />} />

  //             {/* Default */}
  //             <Route path='/' element={<MarvelScreen />} />
  //           </Route>
  //         </Route>
  //       </Routes>
  //     </div>
  //   </Router>
  // );
};

export default AppRouter;
