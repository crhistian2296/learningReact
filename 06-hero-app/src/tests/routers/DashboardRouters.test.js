import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { DashboardRoutes } from '../../routers/DashboardRoutes';

describe('Pruebas en <Dashboard/>', () => {
  const contextValue = {
    user: {
      logged: true,
      name: 'Crhistian',
    },
  };

  test('Debe de renderizar correctamente <DCScreen/>', () => {
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/dc']}>
          <DashboardRoutes />
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.text-info').text().trim()).toBe('Crhistian');
    expect(wrapper.find('h1').text().trim()).toBe('DC Screen');
  });

  test('Debe de renderizar correctamente <SearchScreen/>', () => {
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/search']}>
          <DashboardRoutes />
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.text-info').text().trim()).toBe('Crhistian');
    expect(wrapper.find('h3').at(0).text().trim()).toBe('Search');
  });
});
