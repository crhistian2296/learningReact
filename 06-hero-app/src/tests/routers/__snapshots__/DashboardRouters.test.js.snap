// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Dashboard/> Debe de renderizar correctamente <DCScreen/> 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Asociaciones
      </a>
      <div
        className="navbar-collapse justify-content-sm-between "
      >
        <div
          className="navbar-nav"
        >
          <a
            className="nav-item nav-link"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
          <a
            aria-current="page"
            className="nav-item nav-link active"
            href="/dc"
            onClick={[Function]}
          >
            DC
          </a>
          <a
            className="nav-item nav-link"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
        <div
          className="navbar-nav"
        >
          <span
            className="nav-item nav-link text-info"
          >
            Crhistian
          </span>
          <button
            className="nav-item nav-link btn logoutLink"
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </nav>,
  <div
    className="container my-3"
  >
    <div>
      <h1>
        DC Screen
      </h1>
      <div
        className="row row-cols-xl-3 row-cols-md-2 row-cols-1 g-3 animate__animated animate__fadeIn"
      >
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img-top"
                  src="assets/heroes/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Superman"
                  className="card-img-top"
                  src="assets/heroes/dc-superman.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Superman
                </h5>
                <p
                  className="card-text"
                >
                   
                  Kal-El
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Action Comics #1
                   
                </p>
                <a
                  href="/hero/dc-superman"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Flash"
                  className="card-img-top"
                  src="assets/heroes/dc-flash.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Flash
                </h5>
                <p
                  className="card-text"
                >
                   
                  Jay Garrick
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Jay Garrick, Barry Allen, Wally West, Bart Allen
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Flash Comics #1
                   
                </p>
                <a
                  href="/hero/dc-flash"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Green Lantern"
                  className="card-img-top"
                  src="assets/heroes/dc-green.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Lantern
                </h5>
                <p
                  className="card-text"
                >
                   
                  Alan Scott
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  All-American Comics #16
                   
                </p>
                <a
                  href="/hero/dc-green"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Green Arrow"
                  className="card-img-top"
                  src="assets/heroes/dc-arrow.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Arrow
                </h5>
                <p
                  className="card-text"
                >
                   
                  Oliver Queen
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  More Fun Comics #73
                   
                </p>
                <a
                  href="/hero/dc-arrow"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Wonder Woman"
                  className="card-img-top"
                  src="assets/heroes/dc-wonder.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Wonder Woman
                </h5>
                <p
                  className="card-text"
                >
                   
                  Princess Diana
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  All Star Comics #8
                   
                </p>
                <a
                  href="/hero/dc-wonder"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Martian Manhunter"
                  className="card-img-top"
                  src="assets/heroes/dc-martian.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Martian Manhunter
                </h5>
                <p
                  className="card-text"
                >
                   
                  J'onn J'onzz
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Martian Manhunter
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Detective Comics #225
                   
                </p>
                <a
                  href="/hero/dc-martian"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Robin/Nightwing"
                  className="card-img-top"
                  src="assets/heroes/dc-robin.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Robin/Nightwing
                </h5>
                <p
                  className="card-text"
                >
                   
                  Dick Grayson
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Detective Comics #38
                   
                </p>
                <a
                  href="/hero/dc-robin"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Blue Beetle"
                  className="card-img-top"
                  src="assets/heroes/dc-blue.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Blue Beetle
                </h5>
                <p
                  className="card-text"
                >
                   
                  Dan Garret
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Dan Garret, Ted Kord, Jaime Reyes
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Mystery Men Comics #1
                   
                </p>
                <a
                  href="/hero/dc-blue"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Black Canary"
                  className="card-img-top"
                  src="assets/heroes/dc-black.jpg"
                />
              </div>
              <div
                className="col-8 card-body"
              >
                <h5
                  className="card-title"
                >
                  Black Canary
                </h5>
                <p
                  className="card-text"
                >
                   
                  Dinah Drake
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Dinah Drake, Dinah Lance
                   
                </p>
                <p
                  className="text-muted"
                >
                   
                  Flash Comics #86
                   
                </p>
                <a
                  href="/hero/dc-black"
                  onClick={[Function]}
                >
                  More..
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <Dashboard/> Debe de renderizar correctamente <SearchScreen/> 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Asociaciones
      </a>
      <div
        className="navbar-collapse justify-content-sm-between "
      >
        <div
          className="navbar-nav"
        >
          <a
            className="nav-item nav-link"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
          <a
            className="nav-item nav-link"
            href="/dc"
            onClick={[Function]}
          >
            DC
          </a>
          <a
            aria-current="page"
            className="nav-item nav-link active"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
        <div
          className="navbar-nav"
        >
          <span
            className="nav-item nav-link text-info"
          >
            Crhistian
          </span>
          <button
            className="nav-item nav-link btn logoutLink"
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </nav>,
  <div
    className="container my-3"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-5 mb-4"
      >
        <h3>
          Search
        </h3>
        <hr />
        <form
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Ej: Flash"
            type="text"
            value=""
          />
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Enviar
          </button>
        </form>
      </div>
      <div
        className="col-sm-7"
      >
        <h3>
          Results
        </h3>
        <hr />
        <div
          className="alert alert-info animate__animated animate__fadeIn"
        >
          Search your hero
        </div>
      </div>
    </div>
  </div>,
]
`;
